### PR TITLE
fix: improved serializer selection

### LINKED ISSUE #589: Mocked dynamic response content type not defined in OAS spec
**Describe the bug**
Prism mocks dynamic XML responses from the browser. XML is not defined in my spec.

**To Reproduce**

1. Given this OpenAPI document'

```yaml
openapi: 3.0.0
info:
  title: Example API
  version: '1.0'
  contact:
    name: Bear
    url: 'https://bear.dojo'
    email: pooh@bear.dojo
  termsOfService: 'https://bear.dojo/terms'
  description: The Bear Dojo API
paths:
  /honey:
    get:
      summary: Get Raw Honey
      responses:
        '200':
          description: 'Success, found the bears can have some honey.'
          content:
            application/json:
              schema:
                type: object
                properties:
                  ok:
                    type: boolean
                    example: true
                  honey:
                    type: object
                    properties: 
                       name:
                         type: string
                       location:
                         type: string
                       species:
                         type: string
                         example: killer bees
                required:
                  - ok
                  - honey
      description: Get some raw honey.
      operationId: getRawHoney
```

2. Run this CLI command

`prism mock https://gist.githubusercontent.com/pytlesk4/ddbd7c6f791e68b2b34b0cbd6b11f250/raw/5170c01992bc43c8f8f335ec64ae5ca73257e6e8/bear.dojo.v1.yaml`

3. Open any browser and go to `http://localhost:4010/honey`.

xml-response

**Expected behavior**

Dynamic JSON response.

**Environment (remove any that are not applicable):**
 - Library version: 3.1.0
 - OS: OSX
 - Browser: Firefox

### PR DESCRIPTION
**Pull request by StefanDywersant**
Closes #589

## Checklist

- [x] Tests have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## What kind of change does this PR introduce?

Bugfix

### Current behavior:

`Accept` header negotiation is done in two places: in negotiator and in http-server.
- Negotiatior is selecting matching content based on Accept header and available content types.
- http-server serializes content based on Accept header and available serializers.

To reproduce:
1) spec file has content for application/json
2) request has Accept: application/xml, application/json

Results in:
Negotiator selects application/json as it matches available contents.
http-server serializes content as xml because Accept header has xml and xml serializer exists.

### New behavior:
http-server serialzes the content according to what negotiator negotiated.

## Does this PR introduce a breaking change?

No

### CODE REVIEW

File: packages/http-server/src/server.ts [hunk 1]
```diff
@@ -41,10 +37,11 @@ export const createServer = (operations: IHttpOperation[], opts: IPrismHttpServe
   });
 
   server.addHook('onSend', (request, reply, payload, done) => {
-    if (request.headers.accept === '*/*' && typeof payload !== 'string') {
```
Comment (StefanDywersant) [inline packages/http-server/src/server.ts]: @XVincentX what was the reason behind this `*/*`?

File: packages/http-server/src/server.ts [hunk 2]
```diff
@@ -20,9 +18,7 @@ export const createServer = (operations: IHttpOperation[], opts: IPrismHttpServe
     logger: (components && components.logger) || createLogger('HTTP SERVER'),
     disableRequestLogging: true,
     modifyCoreObjects: false,
-  })
-    .register(formbodyParser)
-    .register(fastifyAcceptsSerializer, { serializers });
```
Comment (StefanDywersant) [inline packages/http-server/src/server.ts]: @XVincentX why selection of serializer was duplicated? (here and as a hook)

**Comment by lag-of-death** [comment]
hi @StefanDywersant, please take a look at https://github.com/stoplightio/prism/pull/604#pullrequestreview-289199069

**Comment by lag-of-death** [comment]
@StefanDywersant, I'll take a look when the conflict is resolved :+1:

File: packages/http-server/src/serialize.ts [hunk 3]
```diff
@@ -33,3 +33,20 @@ export default [
     },
   },
 ];
+
+export const serialize = (payload: unknown, contentType?: string) => {
```
Comment (lag-of-death) [inline packages/http-server/src/serialize.ts]: This function does 3 things:
1) return `undefined`
2) throws an exception
3) serializes with a found serializer

If we need to find a serializer, let's use `Option` at least. Let's not throw when a serializer is not found.
Also, is it possible to do `reply.send(undefined)`?

@XVincentX, WDYT?

File: packages/http-server/src/server.ts [hunk 4]
```diff
@@ -20,9 +18,9 @@ export const createServer = (operations: IHttpOperation[], opts: IPrismHttpServe
...
+  }).register(formbodyParser);
+
+  server.setReplySerializer(payload => payload as string);
```
Comment (XVincentX) [inline packages/http-server/src/server.ts]: This line can be removed if we set the serialiser in the response. See the other comment for other detail.

File: packages/http-server/src/server.ts [hunk 5]
```diff
@@ -119,7 +104,8 @@ export const createServer = (operations: IHttpOperation[], opts: IPrismHttpServe
           if (output.headers) {
             reply.headers(output.headers);
           }
-          reply.send(output.body);
+
+          reply.send(serialize(output.body, reply.getHeader('content-type')));
```
Comment (XVincentX) [inline packages/http-server/src/server.ts]: If possible let's try to keep serialisation and sending separated; this will allow Fastify to give us more precise errors in case something goes wrong.

```suggestion
          reply
            .serializer((payload: unknown) => serialize(payload, reply.getHeader('content-type')))
            .send(output.body);
```

File: test-harness/specs/E2E-Accept-Content-Negotiation-AC-1.oas2.txt [hunk 6]
```diff
@@ -22,7 +22,7 @@ mock -p 4010
 curl -i -X GET http://localhost:4010/todos -H "accept: application/json"
 ====expect====
 HTTP/1.1 200 OK
-content-type: application/json
+content-type: application/json; charset=utf-8
```
Comment (XVincentX) [inline test-harness/specs/E2E-Accept-Content-Negotiation-AC-1.oas2.txt]: This should not be modified. The OAS document does not specify the charset, and so the response should not.

File: packages/http-server/src/serialize.ts [hunk 7]
```diff
@@ -3,10 +3,10 @@ import typeIs = require('type-is');
 
 const xmlSerializer = new j2xParser({});
 
-export default [
+const serializers = [
   {
     regex: {
-      test: (value: string) => !!typeIs.is(value, ['application/*+json', 'application/json']),
+      test: (value: string) => !!typeIs.is(value, ['application/json', 'application/*+json']),
```
[the remark this answers is not in the document]
Comment (StefanDywersant) [reply packages/http-server/src/serialize.ts]: Hm, actually there was no reason. I added it on the front and in the meantime, someone added it on the back and there was a conflict during merging master to my PR.

[the remark this answers is not in the document]
Comment (XVincentX) [reply packages/http-server/src/serialize.ts]: Ah ok no problem, I do not mind, I was just curious.

File: CHANGELOG.md [hunk 8]
```diff
@@ -13,6 +13,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 - Prism is now giving precedence to `application/json` instead of using it as a "fallback" serializer, fixing some conditions where it wouldn't get triggered correctly. [#604](https://github.com/stoplightio/prism/pulls/604)
 - Prism is now taking in consideration the `required` properties for combined schemas (`oneOf, allOf`). This is coming through an update to the Json Schema Faker Library [#623](https://github.com/stoplightio/prism/pulls/623)
 - Prism will never have enough information to return a `403` status code; all these occurences have been now replaced with a `401` status code which is more appropriate [#625](https://github.com/stoplightio/prism/pulls/625)
+- Prism is now selecting proper serializer when Accept header contains content type which is missing in spec. This is a result of simplifying serializer selection approach. [#589](https://github.com/stoplightio/prism/issues/589)
```
[the remark this answers is not in the document]
Comment (StefanDywersant) [reply CHANGELOG.md]: Hm, it's changelog so it's about the changes. And PRs are the way we are performing changes in the code so I'd stay with the current solution. Issues are not directly connected to changelog. At least that's how I understand that ;)

File: packages/http-server/src/serialize.ts [hunk 9]
```diff
@@ -33,3 +33,20 @@ export default [
...
+    return;
+  }
+
+  const serializer = contentType ? serializers.find(s => s.regex.test(contentType)) : undefined;
+
+  if (!serializer) {
+    const error: Error & { status?: number } = new Error(`Cannot find serializer for ${contentType}`);
+    error.status = 500;
+    Error.captureStackTrace(error);
+    throw error;
```
[the remark this answers is not in the document]
Comment (StefanDywersant) [reply packages/http-server/src/serialize.ts]: Done 👍

**Comment by lag-of-death** [comment]
@XVincentX, I understand your reasoning in https://github.com/stoplightio/prism/pull/620#discussion_r325684332

Personally, I would only use the imperative style if something is not possible to do in FP. We don't really need the mayhem of mutations and exceptions being thrown throughout the code. I like your distinction here, it's like Elm/Haskell purity of the code vs side effects taken care of by the compilers, but would try to make that imperative world as small as possible still.

File: packages/http-server/src/__tests__/server.oas.spec.ts [hunk 10]
```diff
@@ -368,3 +368,28 @@ describe.each([['petstore.no-auth.oas2.yaml', 'petstore.no-auth.oas3.yaml']])('s
...
+  it('returns payload matching content-type when accept header is complex', async () => {
+    const response = await server.fastify.inject({
+      method: 'GET',
+      url: '/honey',
+      headers: {
+        accept:
+          'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3',
+      },
+    });
+
```
[the remark this answers is not in the document]
Comment (StefanDywersant) [reply packages/http-server/src/__tests__/server.oas.spec.ts]: Moved, I guess I should add some test case in rails, right?

File: packages/http-server/src/server.ts (hunk 1 above)
Comment (StefanDywersant) [reply packages/http-server/src/server.ts]: Let's assume that edge case is covered with thorough tests :]

File: packages/http-server/src/serialize.ts (hunk 3 above)
Comment (StefanDywersant) [reply packages/http-server/src/serialize.ts]: @lag-of-death gave me a nice lesson on what are those tiny FP pieces about.

However I feel like it is not only about that function, server.ts should be also changed to be compliant with FP approach for the whole thing to make sense (because of throws and nulls).

File: packages/http-server/src/__tests__/server.oas.spec.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (lag-of-death) [reply packages/http-server/src/__tests__/server.oas.spec.ts]: I don't believe we do that

File: CHANGELOG.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (StefanDywersant) [reply CHANGELOG.md]: I updated changelog to point to PR.

File: test-harness/specs/E2E-Accept-Content-Negotiation.complex.oas3.txt [hunk 11]
```diff
@@ -0,0 +1,35 @@
+====test====
+When I send a request to an operation with Accept header containing multiple types.
+And this operation can produce a response that matches nth+1 content type
```
Comment (lag-of-death) [inline test-harness/specs/E2E-Accept-Content-Negotiation.complex.oas3.txt]: hm.. isn't this operation only producing `application/json`?

File: test-harness/specs/E2E-Accept-Content-Negotiation-AC-1.oas2.txt (hunk 6 above)
Comment (StefanDywersant) [reply test-harness/specs/E2E-Accept-Content-Negotiation-AC-1.oas2.txt]: Updating https://github.com/stoplightio/prism/pull/620#discussion_r325693751 resulted in removing all `; charset=utf-8` from tests. Looks like an unexpected fix.

File: packages/http-server/src/server.ts (hunk 5 above)
Comment (StefanDywersant) [reply packages/http-server/src/server.ts]: Done 👍

File: packages/http-server/src/server.ts (hunk 4 above)
Comment (StefanDywersant) [reply packages/http-server/src/server.ts]: Done 👍

File: packages/http-server/src/server.ts (hunk 1 above)
Comment (XVincentX) [reply packages/http-server/src/server.ts]: It was, for sure. We're good here.

File: packages/http-server/src/serialize.ts (hunk 3 above)
Comment (XVincentX) [reply packages/http-server/src/serialize.ts]: Correct, refactoring this is not a 5 minute task. You'll get you hands dirty soon, so I'm not concerned about it. We're good.

File: packages/http-server/src/__tests__/server.oas.spec.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (XVincentX) [reply packages/http-server/src/__tests__/server.oas.spec.ts]: No and hopefully we will never.

File: test-harness/specs/E2E-Accept-Content-Negotiation.complex.oas3.txt (hunk 11 above)
Comment (XVincentX) [reply test-harness/specs/E2E-Accept-Content-Negotiation.complex.oas3.txt]: Yeah indeed, this text is not exactly clear to me

Comment (StefanDywersant) [reply test-harness/specs/E2E-Accept-Content-Negotiation.complex.oas3.txt]: Updated description to be more elaborative. Is it ok now?

Comment (XVincentX) [reply test-harness/specs/E2E-Accept-Content-Negotiation.complex.oas3.txt]: For sure, it's way better now, thanks!
